ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DeprecatedStatementsChecker.java

```diff
@@ -0,0 +1,118 @@
+import java.util.Objects;
+import java.util.Optional;
+
+public class DeprecatedStatementsChecker {
+  public enum Deprecations {
+    DEPRECATED_STREAM_STREAM_JOIN_WITH_NO_GRACE(
+        "DEPRECATION NOTICE: Stream-stream joins statements without a GRACE PERIOD "
+            + "will not be accepted in a future ksqlDB version.\n"
+            + "Please use the GRACE PERIOD clause as specified in "
+            + "https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/"
+            + "select-push-query/"
+    );
+
+    private final String noticeMessage;
+
+    Deprecations(final String noticeMessage) {
+      this.noticeMessage = noticeMessage;
+    }
+
+    public String getNoticeMessage() {
+      return noticeMessage;
+    }
+  }
+
+  final MetaStore metaStore;
+
+  DeprecatedStatementsChecker(final MetaStore metaStore) {
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+  }
+
+  public Optional<Deprecations> checkStatement(final Statement statement) {
+    if (isStreamStreamJoinWithoutGraceStatement(statement)) {
+      return Optional.of(Deprecations.DEPRECATED_STREAM_STREAM_JOIN_WITH_NO_GRACE);
+    }
+
+    return Optional.empty();
+  }
+
+  private boolean isStreamStreamJoinWithoutGraceStatement(final Statement statement) {
+    if (statement instanceof CreateAsSelect) {
+      return isStreamStreamJoinWithoutGraceQuery(((CreateAsSelect) statement).getQuery());
+    } else if (statement instanceof Query) {
+      return isStreamStreamJoinWithoutGraceQuery((Query) statement);
+    }
+
+    return false;
+  }
+
+  private boolean isStreamStreamJoinWithoutGraceQuery(final Query query) {
+    if (query.getFrom() instanceof Join) {
+      final Join join = (Join) query.getFrom();
+
+      // check left joined source is a stream
+      if (!isJoinedSourceStream(join.getLeft())) {
+        return false;
+      }
+
+      for (final JoinedSource joinedSource : join.getRights()) {
+        // check right joined source is a stream
+        if (!isJoinedSourceStream(joinedSource.getRelation())) {
+          break;
+        }
+
+        if (!joinedSource.getWithinExpression().flatMap(WithinExpression::getGrace).isPresent()) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private boolean isJoinedSourceStream(final Relation relation) {
+    // DataSourceExtractor must be initialized everytime we need to extract data from a node
+    // because it changes its internal state to keep all sources found
+    final DataSourceExtractor dataSourceExtractor = new DataSourceExtractor(metaStore, false);
+
+    // The relation object should have only one joined source, but the extract sources method
+    // returns a list. This loop checks all returned values are a Stream just to prevent throwing
+    // an exception if more than one value is returned.
+    for (final AliasedDataSource source : dataSourceExtractor.extractDataSources(relation)) {
```

Comment:
I agree to the sentiment to not crash. But I am wondering if we should log a WARN if there is more than one source found?

Reply:
Added